tools/spm/generate_partition_code.py

```diff
@@ -0,0 +1,439 @@
+#!/usr/bin/python
+import glob
+import itertools
+import json
+import os
+from os.path import join as path_join
+
+from jinja2 import Environment, FileSystemLoader
+from jsonschema import validate
+
+SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
+TEMPLATES_DIR = path_join(SCRIPT_DIR, 'templates')
+
+
+def assert_int(num):
+    """
+    Tries to parse an integer num from a given string
+    
+    :param num: Number in int/string type
+    :return: Numeric value
+    """
+    if isinstance(num, int):
+        return num
+    num_str = str(num)
+    radix = 16 if num_str.lower().startswith('0x') else 10
+    res = int(num_str, radix)
+    # Python converts str to int as a signed integer
+    if res > 0x7FFFFFFF:
+        res -= 0x100000000
+    return res
+
+
+class SecureFunction(object):
+    MINOR_POLICIES = ['strict', 'relaxed']
+
+    def __init__(self, sfid, signal, non_secure_clients, minor_version=1, minor_policy='strict'):
+        """
+        Secure Function C'tor (Aligned with json schema)
+        
+        :param sfid: Secure function identifier (available to user)
+        :param signal: Secure Function identifier inside the partition
+        :param non_secure_clients: True to allow connections from non-secure partitions
+        :param minor_version: Secure function version
+        :param minor_policy: Enforcement level of minor version
+        """
+        self.sfid = sfid
+        self.signal = signal
+
+        assert isinstance(non_secure_clients, bool), 'non_secure_clients parameter must be of boolean type'
+        self.nspe_callable = non_secure_clients
+
+        self.minor_version = assert_int(minor_version)
+        assert self.minor_version > 0, 'minor_version parameter is invalid'
+
+        assert minor_policy in self.MINOR_POLICIES, 'minor_policy parameter is invalid'
+        self.minor_policy = minor_policy
+
+    def __eq__(self, other):
+        return (
+            (self.sfid == other.sfid) and
+            (self.signal == other.signal) and
+            (self.nspe_callable == other.nspe_callable) and
+            (self.minor_version == other.minor_version) and
+            (self.minor_policy == other.minor_policy)
+        )
+
+
+class MmioRegion(object):
+    MMIO_PERMISIONS = {'READ-ONLY': 'PSA_MMIO_PERM_READ_ONLY', 'READ-WRITE': 'PSA_MMIO_PERM_READ_WRITE'}
+
+    def __init__(self, **kwargs):
```

Comment:
This constructor seems to take:
 * `permission`
 * `base` (optional)
 * `size` (optional)
 * `name` (optional)

Could you write it as:
```python
def __init__(self, permission, base=None, size=None, name=None):
    """Docstring
    ...
    """
    assert (base and size) or name
    self.permission = self.MMIO_PERMISIONS[kwargs['permission']]
    if 'name' in kwargs:
       self.base = kwargs['name']
       self.size = '(sizeof(*({})))'.format(kwargs['name']
    if 'base' in kwargs:
       self.base = kwargs['base']
       self.size = assert_int(kwargs['size'])
```
Do you see a large downside to writing it that way?